torch/_prims/utils.py

```diff
@@ -0,0 +1,468 @@
+#   - a non-scalar or non-Tensor is given
+#   - the shape of any tensors is distinct
+def check_same_shape(*args):
+    """
+    Checks that all Tensors in args have the same shape.
+
+    Raises a RuntimeError when:
+      - args contains an object whose type is not Tensor or Number
+      - two Tensor objects in args have different devices
+    """
+    shape = None
+
+    for arg in args:
+        if isinstance(arg, Number):
+            continue
+        elif isinstance(arg, TensorLike):
+            if shape is None:
+                shape = arg.shape
+
+            if not is_same_shape(shape, arg.shape):
+                msg = "Shape {0} is not the expected shape {1}!".format(
+                    arg.shape, shape
+                )
+                raise RuntimeError(msg)
+        else:
+            msg = (
+                "Unexpected type when checking for same shape, " + str(type(arg)) + "!"
+            )
+            raise RuntimeError(msg)
+
+
+_integer_dtypes = (torch.uint8, torch.int8, torch.int16, torch.int32, torch.int64)
+_float_dtypes = (torch.float16, torch.bfloat16, torch.float32, torch.float64)
+_complex_dtypes = (torch.complex64, torch.complex128)
+
+
+def is_boolean_dtype(dtype: torch.dtype) -> bool:
+    return dtype is torch.bool
+
+
+def is_integer_dtype(dtype: torch.dtype) -> bool:
+    return dtype in _integer_dtypes
+
+
+def is_float_dtype(dtype: torch.dtype) -> bool:
+    return dtype in _float_dtypes
+
+
+def is_complex_dtype(dtype: torch.dtype) -> bool:
+    return dtype in _complex_dtypes
+
+
+def dtype_to_type(dtype: torch.dtype) -> type:
+    """
+    Computes the corresponding Python type (AKA "type kind") for the
+    given dtype.
+    """
+    assert isinstance(dtype, torch.dtype)
+
+    if dtype is torch.bool:
+        return bool
+    if dtype in _integer_dtypes:
+        return int
+    if dtype in _float_dtypes:
+        return float
+    if dtype in _complex_dtypes:
+        return complex
+
+    raise ValueError("Invalid dtype!")
+
+
+_type_to_dtype_map = {
+    bool: torch.bool,
+    int: torch.int64,
+    float: torch.float64,
+    complex: torch.complex128,
+}
+
+
+def type_to_dtype(typ: type) -> torch.dtype:
+    """
+    Computes the corresponding dtype for a Number type.
+    """
+    return _type_to_dtype_map[typ]
+
+
+_ordered_types = (bool, int, float, complex)
+
+
+def get_higher_type(a: type, b: type) -> type:
+    """
+    Returns the higher of the two given Number types.
+
+    The types are ordered bool -> int -> float -> complex.
+    """
+    # Type checking
+    assert a in _ordered_types
+    assert b in _ordered_types
+
+    if a is b:
+        return a
+
+    for typ in _ordered_types:
+        if a is typ:
+            return b
+        if b is typ:
+            return a
+
+    raise ValueError("Unknown Python scalar type!")
+
+
+# Returns the higher of two torch datatypes a and b or, if the two
+#   are not ordered relative to each other, the next
+#   higher datatype
```

Comment:
I would love to see the word "lattice" here :P

Reply:
Sure